x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java

```diff
@@ -211,12 +212,15 @@ void validateIndexPermissionsAreSubset(
      * @param authorizationInfo information used for authorization, for a specific Subject, that was previously retrieved
      *                          using {@link #resolveAuthorizationInfo(Subject, ActionListener)}
      * @param privilegesToCheck the object that contains the privileges to check for the Subject
+     * @param runDetailedCheck whether to return the details on which privileges are granted and which are not.
+     *                         {@code PrivilegesCheckResult#details} must be {@code null} only if this parameter is {@code false}.
      * @param applicationPrivilegeDescriptors a collection of application privilege descriptors
      * @param listener the listener to be notified of the check privileges response
      */
     void checkPrivileges(
         AuthorizationInfo authorizationInfo,
         PrivilegesToCheck privilegesToCheck,
+        boolean runDetailedCheck,
```

Comment:
I understand this PR is still draft, but I strongly think `runDetailedCheck` should be part of `PrivilegesToCheck` class instead of a separate parameter. A few reasons:
* In a single request flow, there is always just one way to check the privileges, either complete or short-circuited. It is by nature part of the check request itself, not a separate thing. I don't think we would ever support running both complete and short-circuited checks for a single request.
* The values for `privilegesToCheck` and `runDetailedCheck` should be determined at REST layer. We talked about that in future ProfileHasPrivileges can also run the complete check on an optional parameter. This parameter is from the REST layer and should be part of the privilegeCheck request itself. 
* Conceptually, a 1-to-1 relationship between check and result is much easier to understand. So I'd keep the variation of the check inside the check itself instead of keeping it outside as a separate variable.

Reply:
Thanks for taking my suggestion. I really appreciate it. Here are some clarifications on my reasonings for posterity.

> I backtracked on it because the PrivilegesToCheck existing fields are parsed directly from the request body (for the "profile has privilege" endpoint) and it felt hacky to sneak in an extra argument in the constructor there.

My thinking is that the `mode` parameter may as well also be parsed from the request body. For now, we will supply it to the parsing method. But when we support optional full check for profileHasPrivilege in future, this `mode` will be something that users provide in the request body.

> I actually feel differently on this. I see it as a single "collection" of privileges that can be checked in 2 different ways. The class is used to group the "things" to check and the new parameter is about the "mode" to check.

I agree your way of viewing it as 'a single "collection" of privileges that can be checked in 2 different ways'. But my conclusion is different (in whether they should be kept together). My thinking is that there should be a single class that represent the entire check request. Maybe our different views have something to do with the class name. Imagine if it is called HasPrivilegeRequest (not proposing this name), I guess it would feel more natural to have it enclose both pieces of information?